### PR TITLE
Add New Material Design Icons to its own region

### LINKED ISSUE #279: Material Design Icons missing some icons
#### 🎯 Subject of the issue

It appears that some of the icons from the Material Design Icons set are missing, the ones I've found to be missing are:
* `battery-charging-outline`
* `battery-charging-10`
* `battery-charging-50`
The icons are not listed in the cheat sheet on [the nerd-fonts website](http://nerdfonts.com/#cheat-sheet) but they are on [the MDI website](https://materialdesignicons.com/) if you search for 'charging'

#### 🔧 Your Setup

- _Which font are you using (e.g. `Anonymice Powerline Nerd Font Complete.ttf`)?_ 
Sauce Code Pro Nerd Font Complete.ttf
- _Which terminal emulator are you using (e.g. `iterm2`, `urxvt`, `gnome`, `konsole`)?_ 
st
- _Are you using OS X, Linux or Windows? And which specific version or distribution?_ 
Debian 9 Testing

#### ★ Optional
As you can see there are missing icons here
Missing Icons on website

Here are what all the charging icons should be
All the correct icons

### LINKED ISSUE #365: [Suggestion] Fix invalid code points for some glyphs
## Summary

The current builds overwrites some code points that Unicode Consorthium prohibits to use for custom glyphs. I want to change this by fixing font_patcher.

## Problem Detail

Unicode defines [Private Use Areas][pua] and the consorthium itself decide not to add characters on these areas. So we can add glyphs as we like.

[pua]: https://en.wikipedia.org/wiki/Private_Use_Areas

The areas has these code points: U+E000..U+F8FF, U+F0000..U+FFFFD, U+100000..U+10FFFD.

But the Nerd Fonts is overwriting more code points than them. The font_patcher writes glyphs as below.

* source: the ones from source font files.
* current: the ones font_patcher writes into.

| Font Name               |   source  |    current    |   plan 1  |          plan 2         |
|-------------------------|:---------:|:-------------:|:---------:|:-----------------------:|
| Seti-UI + Custom        | E4FA-E52E |   E5FA-E62E   |     ←     |            ←            |
| Devicons                | E600-E6C5 |   E700-E7C5   | E630-E6F6 | E700-E7C5 (not changed) |
| Powerline Symbols       | E0A0-E0B3 |       ←       |     ←     |            ←            |
| Powerline Extra Symbols | E0A3-E0D4 |       ←       |     ←     |            ←            |
| Pomicons                | E000-E00A |       ←       |     ←     |            ←            |
| Font Awesome            | F000-F2E0 |       ←       |     ←     |            ←            |
| Font Awesome Extension  | E000-E0A9 |   E200-E2A9   |     ←     |            ←            |
| Power Symbols           | 23FB-2B58 |       ←       |     ←     |            ←            |
| Material                | F001-F847 | **F500-FD46** | E700-EF47 |   F500-F8FF,E800-EC47   |
| Weather Icons           | F000-F0EB |   E300-E3EB   |     ←     |            ←            |
| Font Logos (Font Linux) | F100-F11C |   F300-F31C   |     ←     |            ←            |
| Octicons                | F000-F105 |   F400-F505   |     ←     |            ←            |
| Octicons                | 2665-2665 |       ←       |     ←     |            ←            |
| Octicons                | 26A1-26A1 |       ←       |     ←     |            ←            |
| Octicons                | F27C-F27C |   F4A9-F4A9   |     ←     |            ←            |

It is the problem that font_patcher writes glyphs from Material into the range: **U+F500 - U+FD46**. This range overlaps areas that should not use for such purpose.

| area           | name                                                                                         |
|----------------|----------------------------------------------------------------------------------------------|
| U+F900..U+FAFF | [CJK Compatibility Ideographs](https://en.wikipedia.org/wiki/CJK_Compatibility_Ideographs)   |
| U+FB00..U+FB4F | [Alphabetic Presentation Forms](https://en.wikipedia.org/wiki/Alphabetic_Presentation_Forms) |
| U+FB50..U+FDFF | [Arabic Presentation Forms-A](https://en.wikipedia.org/wiki/Arabic_Presentation_Forms-A)     |

## Suggestion

So I suggest two plans to solve this.

### plan 1

1. Move Devicons just after Seti-UI + Custom.
2. Move Material to U+E700..U+EF47.

* Pros - Material will be still in a cluster.
* Cons - Two glyph sets will be moved.

### plan 2

1. Move and separate Material into U+F500..U+F8FF, U+E800-U+EC47.

* Pros - Only one set will be moved.
* Pros - The former part of Material still has the same code points.
* Cons - Material will be separated into two clusters.

----

I prefer _plan 2_ because it has less impact on the current builds. How do you think?

### PR DESCRIPTION
**Pull request by earboxer**
#### Description

This is an alternative to #772 . It's part of "Plan 5" Plus in #365. (another part would be to remove the old icons when we make the breaking change).

#### Requirements / Checklist

- [X] Read the [Contributing Guidelines](https://github.com/ryanoasis/nerd-fonts/blob/master/contributing.md)
- [X] Verified the license of any newly added font, glyph, or glyph set

#### What does this Pull Request (PR) do?

Add the updated material design icons in its canonical location.

#### What are the relevant tickets (if any)?

Fixes #365 

and Fixes #279 

But it doesn't make any breaking changes, so this can be merged before the 3.0 release

### CODE REVIEW

**Comment by Finii** [comment]
I am always a bit in fear when fonts originate from webfont people. As they often do not care about codepoints and codepoint consistency (they reference glyphs by name, rather).

I examined 3 releases: 3.4.93, 4.6.95, 6.1.95 and they all differ slightly (apart from the switch to `F0001`), the worst is the drop of one glyph with renumbering all following ones, is that what you noticed? Between 4.x and 6.x

image

And then individual glyphs are replaced all over now and then, _Notes_ become _Cooking Pot_, _No Entry_ becomes _No PianoKeys_, _Flower Pot?_ becomes _Flag With Star_ ... and that is in all releases.

So ... is this really a basis to use?

**Comment by Finii** [comment]
Maybe I'm too skeptic with the random changes in the comment above. I could not find any changes in more recent versions. Possibly they underwent some restructuring.

What I do not like too much, but that is personal, is that the svg-to-ttf is hidden in a npm thing, but I wanted to check how they keep the glyph codepoints constant. But when I read npm, my eyes glaze over and I suddenly loose all interest.)

File: readme.md [hunk 1]
```diff
@@ -89,7 +89,7 @@ _If you..._
   * For more details see the [**Font Patcher**](#font-patcher) section
 * **`51`** already [patched font families](#patched-fonts)
 * Over **`1,444,400`** unique combinations/variations of patched fonts [(more details)](#combinations)
-* Over **`2,824`** glyphs/icons combined [(more details)](#combinations)
+* Over **`9,000`** glyphs/icons combined [(more details)](#combinations)
```
[the remark this answers is not in the document]
Comment (earboxer) [reply readme.md:92]: > I can hardly believe that someone needs 'sim card 2 has connection problems' glyphs are needed in a powerline setup or while programming in an editor ;-)

Hmm, that's interesting. I'm working on [this](https://todo.sr.ht/~mil/sxmo-tickets/426) as a contributor to [SXMO](https://sxmo.org), a mobile phone UI, that, among other things, allows you to send text messages from an SSH session.

Really, there's 4 places we render the icons:
* in the status bar (swaybar in sway, xsetroot in dwm?)
* in dmenu/bemenu
* In a local (on-device) terminal
* in a remote terminal

...

It sounds like we're probably gonna move away from nerdfonts anyway, so, thanks for the tip about font fallback!

**Comment by earboxer** [comment]
> I wanted to check how they keep the glyph codepoints constant

MaterialDesign has a `meta.json` (similar format that a lot of the FontAwesome-derivatives have).

With a little jq-foo:

`jq 'map([.name,.codepoint])' <meta.json`

But it seems like they manage their repo using external tools, so, yeah, valid concerns.

**Comment by Finii** [comment]
Scrolling through all the glyphs...
I believe that it needs to utilize some `ScaleGlyph` stuff, or some pairs will become unusable.
Yes, the current one also misses it.

image
image
image

Maybe one ('old style') `ScaleGlyph` is enough, because all icons maintain a general good same-scale.

File: readme.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (Finii) [reply readme.md:92]: \*laugh\*

Ok, there are uses, obviously. :smile:

Yes, font fallback, if it is possible on your OS... and if the icons do not match the size of the text font (i.e. EM wise), one can still just scale the symbol font in total, very easy, with `fontforge`.

**Comment by Finii** [comment]
* Rebase on master
* Use desktop font instead of webfont
* Create cheat sheet data
* Force push

**Comment by Finii** [comment]
Thanks for this (and the other 772) PR.
Finally we start moving, sorry for the long wait.

There is still some unsolved problem with the `scaleGlyph`. A lot of paired-symbols look bad if you use them together. In principle the `scaleGlyph` should fix it, but ... it has problems. #837 tried to solve it, partially, and the same issue came up in ... <still searching>.

**Comment by Finii** [comment]
Note to self: Is the cheat sheet rendering still ok with 5 digit codes?

**Comment by Finii** [comment]
Rebase on master, force push

**Comment by Finii** [comment]
(As expected) scaling broken, refer same glyphs in [comment](https://github.com/ryanoasis/nerd-fonts/pull/773#issuecomment-1023497439) above

image

Introducing `ScaleGlyph` or `ScaleGroup` lets see

**Comment by Finii** [comment]
Material Design Icons has a bug in one glyph:

image

**Comment by Finii** [comment]
Pulled bugfix [05a9ec8](https://github.com/ryanoasis/nerd-fonts/commit/05a9ec8b6d3ab6ede3e519adc99b1306ef33a663) already to master. Rebase, force push.

**Comment by Finii** [comment]
Rebase on master, force push.
We near us a merge and also a release :tada:

**Comment by Finii** [comment]
This seems to be stuck: https://github.com/Templarian/MaterialDesign-Font/issues/9

So we might need to fix it on our own.

**Pull request merged**